_source/_posts/2022-08-15-spring-boot-crud-update.md

```diff
@@ -0,0 +1,1077 @@
+
+Create a `.env` file in the client project root directory. The **Client ID** and **Issuer URI** are the values you used above in the Spring Boot `application.properties` file. The **Server URI** is the local URI for the Spring Boot server is you can leave as it is unless you made a change (this gets used in the `Api.js` module).
+
+`.env`
+
+```env
+VUE_APP_CLIENT_ID=<your-client-id>
+VUE_APP_ISSUER_URI=<your-issuer-uri>
+VUE_APP_SERVER_URI=http://localhost:9000
+```
+
+It's important to note that putting values like this in a `.env` file in a client application does not make then secure. It helps by keeping them out of a repository. However, they are still public in the sense that they are necessarily visible in the Javascript code sent to the browser. In this use case, it's more of a configuration and organizational tool than a security tool. 
+
+Replace `App.vue` with the following.
+
+`src/App.vue`
+
+{% raw %}
+```vue
+<template>
+  <q-layout view="hHh lpR fFf">
+
+    <q-header elevated class="bg-primary text-white">
+      <q-toolbar>
+        <q-toolbar-title>
+          <q-avatar>
+            <q-icon name="kayaking" size="30px"></q-icon>
+          </q-avatar>
+          Todo App
+        </q-toolbar-title>
+        {{this.claims && this.claims.email ? claims.email : '' }}
+        <q-btn flat round dense icon="logout" v-if='authState && authState.isAuthenticated' @click="logout"/>
+        <q-btn flat round dense icon="account_circle" v-else @click="login"/>
+      </q-toolbar>
+    </q-header>
+
+    <q-page-container>
+      <router-view></router-view>
+    </q-page-container>
+
+  </q-layout>
+</template>
+
+<script>
+export default {
+  name: 'LayoutDefault',
+  data: function () {
+    return {
+      claims: null
+    }
+  },
+  watch: {
+    'authState.isAuthenticated'() {
+      this.$log.debug(("watch triggered!"))
+      this.updateClaims()
+    }
+  },
+  created() {
+    this.updateClaims()
+  },
+  methods: {
+    async updateClaims() {
+      if (this.authState && this.authState.isAuthenticated) {
+        this.claims = await this.$auth.getUser()
+      }
+    },
+    async login() {
+      await this.$auth.signInWithRedirect({originalUri: '/todos'})
+    },
+    async logout() {
+      await this.$auth.signOut()
+    }
+  },
+}
+</script>
+```
+{% endraw %}
+
+This is the top-level component that defines the header bar and includes the router component. The header bar has a login or logout button and will show the authenticated user's email address when logged in.
+
+The app gets the authenticated user's email address from the JWT claims (a claim is a piece of information asserted about the subject by the authenticating authority). This happens in the `updateClaims()` method, which is triggered when the component is created and is also triggered by a watch method so that it is updated as the authenticated state changes.
```

Comment:
```suggestion
The app gets the authenticated user's email address from the JWT claims. (A claim is a piece of information asserted about the subject by the authenticating authority.) This happens in the `updateClaims()` method, which is triggered when the component is created, and is also triggered by a watch method so that it is updated as the authenticated state changes.
```